File: typescript/libs/system/workspace/domain/src/lib/istrategy.ts
import {
  AssociationType,
  Class,
  MethodType,
  RoleType,
} from '@allors/system/workspace/meta';
import { IDiff } from './diff/idiff';
import { IObject } from './iobject';
import { ISession } from './isession';
import { IUnit } from './types';

export interface IStrategy {
  session: ISession;

  object: IObject;

  cls: Class;

  id: number;

  version: number;

  isNew: boolean;

  hasChanges: boolean;

  isDeleted: boolean;

  delete(): void;

  reset(): void;

  diff(): IDiff[];

  canRead(roleType: RoleType): boolean;

  canWrite(roleType: RoleType): boolean;

  canExecute(methodType: MethodType): boolean;

  existRole(roleType: RoleType): boolean;

  hasChanged(roleType: RoleType): boolean;

  restoreRole(assignedRoleType: RoleType): void;

  getRole(roleType: RoleType): unknown;

  getUnitRole(roleType: RoleType): IUnit;

  getCompositeRole<T extends IObject>(
    roleType: RoleType,
    skipMissing?: boolean
  ): T;

  getCompositesRole<T extends IObject>(
    roleType: RoleType,
    skipMissing?: boolean
  ): ReadonlyArray<T>;

  setRole(roleType: RoleType, value: unknown): void;

  setUnitRole(roleType: RoleType, value: IUnit): void;

  setCompositeRole<T extends IObject>(roleType: RoleType, value: T): void;

  setCompositesRole<T extends IObject>(
    roleType: RoleType,
    value: ReadonlyArray<T>
  ): void;

  addCompositesRole<T extends IObject>(roleType: RoleType, value: T): void;

  removeCompositesRole<T extends IObject>(roleType: RoleType, value: T): void;

  removeRole(roleType: RoleType): void;

  getCompositeAssociation<T extends IObject>(
    associationType: AssociationType
  ): T;

  getCompositesAssociation<T extends IObject>(
    associationType: AssociationType
  ): ReadonlyArray<T>;
}

File: dotnet/System/Workspace/Allors.Workspace.Domain/IStrategy.cs
// <copyright file="Object.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace
{
    using System.Collections.Generic;
    using Meta;
    using Signals;

    public interface IStrategy
    {
        ISession Session { get; }

        IObject Object { get; }

        IClass Class { get; }

        long Id { get; }

        ISignal<long> Version { get; }

        ISignal<bool> IsNew { get; }

        ISignal<bool> HasChanges { get; }

        void Reset();

        IReadOnlyList<IDiff> Diff();

        ISignal<bool> CanRead(IRoleType roleType);

        ISignal<bool> CanWrite(IRoleType roleType);

        ISignal<bool> CanExecute(IMethodType methodType);

        ISignal<bool> ExistRole(IRoleType roleType);

        ISignal<bool> HasChanged(IRoleType roleType);

        void RestoreRole(IRoleType roleType);

        IRoleSignal<object> ScalarRole(IRoleType roleType);

        IRoleSignal<T> ScalarRole<T>(IRoleType roleType);

        IRoleSignal<T> CompositeRole<T>(IRoleType roleType) where T : class, IObject;

        ICompositesRoleSignal<T> CompositesRole<T>(IRoleType roleType) where T : class, IObject;

        IDerivedRoleSignal<T> DerivedRole<T>(IRoleType roleType);

        IAssociationSignal<T> CompositeAssociation<T>(IAssociationType associationType) where T : class, IObject;

        ICompositesAssociationSignal<T> CompositesAssociation<T>(IAssociationType associationType) where T : class, IObject;

        IMethodSignal Method(IMethodType methodType);
    }
}
